metals/src/main/scala/scala/meta/internal/metals/JarTopLevels.scala

```diff
@@ -0,0 +1,111 @@
+package scala.meta.internal.metals
+
+import java.nio.file.Files
+import java.nio.file.attribute.BasicFileAttributeView
+import java.sql.{Connection, PreparedStatement, Statement}
+
+import scala.collection.concurrent.TrieMap
+import scala.meta.internal.io.PlatformFileIO
+import scala.meta.internal.metals.JdbcEnrichments._
+import scala.meta.io.AbsolutePath
+
+/**
+ * Handles caching of Jar Top Level Symbols in H2
+ *
+ * Wrapper around the indexed_jar and toplevel_symbol sql tables.
+ */
+final class JarTopLevels(conn: () => Connection) {
+
+  /**
+   * Retrieves top level Scala symbols of a jar from H2
+   *
+   * @param path absolute path of the jar
+   * @param digest MD5 digest of the jar
+   * @return the top level Scala symbols in the jar
+   */
+  def getTopLevels(
+      path: AbsolutePath
+  ): Option[TrieMap[String, AbsolutePath]] = {
+    val fs = PlatformFileIO.newJarFileSystem(path, create = false)
+    conn()
+      .query(
+        "select id from indexed_jar where path=? and last_modified=? and size=?"
+      ) { q =>
+        val attributes = getAttributes(path)
+        q.setString(1, path.toString)
+        q.setLong(2, attributes.lastModifiedTime().toMillis)
+        q.setLong(3, attributes.size())
+      } { _.getInt(1) }
+      .map { jar =>
+        val toplevels = TrieMap[String, AbsolutePath]()
+        conn().query(
+          "select symbol, path from toplevel_symbol where jar=?"
+        )(q => q.setInt(1, jar)) { rs =>
+          toplevels(rs.getString(1)) = AbsolutePath(fs.getPath(rs.getString(2)))
+        }
+        toplevels
+      }
+      .headOption
+  }
+
+  /**
+   * Stores the top level symbols for the Jar in H2
+   *
+   * @param path absolute path of the jar
+   * @param digest MD5 digest of the jar
+   * @param toplevels toplevel symbols in the jar
+   * @return the number of toplevel symbols inserted
+   */
+  def putTopLevels(
+      path: AbsolutePath,
+      toplevels: TrieMap[String, AbsolutePath]
+  ): Int = {
+    // Add jar to H2
+    var jarStmt: PreparedStatement = null
+    val jar = try {
+      jarStmt = conn().prepareStatement(
+        s"insert into indexed_jar (path, last_modified, size) values (?, ?, ?)",
```

Comment:
Do we clean up unused entries? It would be good if we garbage collect dead indexes. It might be simpler to implement that if we store a single id for each jar entry (instead of `path, last_modified, size`), then we could cascade drop entries that are not in the last build import.

Reply:
the single ID can be a concatenation or checksum of the three values

Reply:
So you want me collect all Jars that are part of the build during indexing and then at the end of indexing drop the ones that aren't.

Reply:
I think that would be good if we can implement it.

Reply:
Unused Jars/Toplevels are removed from cache after indexing now.